### ISSUE TITLE
Get task retvals from the nursery after it finishes

### ISSUE THREAD

**Issue by dubslow** [python-trio/trio#2633]
I haven't seen a way to do this in the docs, and it certainly seems very logical/thematic if it were to work like this. The `nursery` is the starter and finisher of tasks, its sole and entire purpose is to start tasks and ensure they finish in some halfway-sensible way, with some sort of sane control flow.

Therefore it's only logical that the return values of those tasks are collected by the nursery. Return values shouldn't disappear into the void any more than errors. The nursery gathers errors, so it should gather retvals too.

I find myself wanting to gather return values from my tasks somewhat akin to the following, taken from the docs: running a process for an arbitrary amount of time which dynamically spawns an arbitrary number of tasks.

A little bit of searching finds me a recipe to kinda back my way into this functionality faking out a sort of `trio.gather`, but it seems to me to be a pretty basic primitive of structured concurrency, and if it's common enough to be a recipe, then it's common enough to implement once and get right instead of rewriting it (or at least repasting and in-place-modifying it) every time you call a nursery.

But then I'm quite new to asyncio and structured concurrency. All I know is that I greatly enjoyed NJS's blog post on control flow, and have equally enjoyed my first 100 lines of code. But I would love even more to either 1) have this primitive implemented in trio 2) understand why it's not a primitive in some way. It sure sounds like one: "nurseries start tasks, nurseries finish tasks, nurseries call functions, nurseries collect those functions' retvals"

**Comment by smurfix** [comment]
> if it's common enough to be a recipe, then it's common enough to implement once and get right

The problem is: What is "get right"? In many cases you want partial results from subtasks as soon as they're available. (Possible solution: Use a memory channel.) Sometimes you also need to associate task names/objects with the results (send a name+result tuple to the channel.) In others you need a list (write a two-liner that appends the result to an array) or associated with an object (store the result in the object, or write another two-liner that saves it to a dict.)  In many others you don't need the return values at all and don't want to save a possibly-unbounded list of `None` values. And sometimes you need to combine any of the above with a concurrency limiter, which also might have to decide what to do about the return value.

The overall idea is that, confronted with the fact that no one-size-fits-all recipe exists (IMHO cannot exist), you think about the problem as you encounter it, and choose the method that best fits your situation instead of trying to shoehorn your problem into the one-size-fits-all recipe.

We probably should add this to the documentation somewhere. Assuming it's not in there already …

**Comment by A5rocks** [comment]
Hi! For your specific problem, I presume something like the pattern mentioned in https://github.com/python-trio/trio/issues/2632#issuecomment-1506837686 works for you?

(Where you just append the result to a list or add it to a dict or something)

If it does, should we just document this particular composition of features? If not, could you provide a more concrete example? I don't think I get what you want.

As for why not just add a kwarg or something, I'm more in favor of just using a pretty basic composition strategy like this cause it's just more flexible! Just like how we don't have timeout args (normally, except for like if there's multiple sub-timeouts such as making an http request) and instead have a context manager!

**Comment by dubslow** [comment]
Disclaimer: am noob, but I don't see how that makes the result available outside the `with` statement. Indeed as I edited into my previous comment, in principle our code may not know the final result *until* the nursery closes. Perhaps I'm not paying enough attention to cancellation semantics, because I suppose I'm saying that when a task is cancelled, it would clean itself up by returning its final result -- where the result isn't final until it's cancelled.

So perhaps I've been not stating my question correctly.

My head hurts lol

**Comment by dubslow** [comment]
Here's a toy example expanded from a throwaway in the docs. Obviously it doesn't run, and as I said in my previous comment I think it essentially doesn't work as I intended w.r.t. exceptions/cancellations, but that's the semantics I'm looking for...

https://gist.github.com/dubslow/449c7743e917b6130719df7be01b8b26

**Comment by dubslow** [comment]
And modifying that, I've come up with a toy example that actually works and does what I want it to:

https://gist.github.com/dubslow/379342c6dbc0efc69491fe9c8becc089

However I find this code to be extremely smelly, and a considerable anti-pattern. I would *love* for a better way to do this.

example output:

```python nursery_retvals.py 
awaiting next connection: asdf
got text='asdf' from new conn
awaiting next connection: gg
got text='gg' from new conn
awaiting next connection: go next
got text='go next' from new conn
awaiting next connection: Hello, world!
got text='Hello, world!' from new conn
awaiting next connection: ^C
got text='' from new conn
After server closed, we've counted lifetime total conns=5 messages=9
```

**Comment by A5rocks** [comment]
Hmm.. I have to say I still don't quite get the problem. Like, this is what I think of when I think "asyncio.gather" to try to demonstrate what I meant by that pattern (I haven't tested this worked, for what it's worth):

```py
import trio

async def gather(tasks):
  results = dict({t: None for t in tasks})  # in-order results
  async def run(task):
    results[task] = await task()

  async with trio.open_nursery() as nursery:
    for task in tasks:
      nursery.start_soon(run, task)

  return results.values()
  ```

Looking at your toy example, I don't quite get how you expect to have the counters done for you? Sorry for any confusion and for taking a while to reply, got very distracted!

**Comment by A5rocks** [comment]
Ah sorry if I'm not really being clear on _why_ I'm saying this:

I think I agree with @smurfix that there are so many specific things you might want (`asyncio.gather` (which has a boolean flag for operation), `asyncio.as_completed`, `asyncio.wait` (which overlaps with `asyncio.gather` with a 3-member enum flag for different types of operation)).

As such, I'd *much* rather people figure out this pretty core pattern and have it down. Going from "oh I need to do some work for every task" or "I need to prepare stuff before every task" or "how can I keep a `outcome.Outcome` for all run tasks" to "oh I can just use a function" is an instinct I'd personally rather users have (disclaimer: I'm not involved in library design :^) and I do think having docs that cover this case would be nice, as I do not remember the docs having something similar.

**Comment by dolamroth** [comment]
I think, you may just use https://github.com/tiangolo/asyncer It is built with anyio, so it's trio-compatible
https://asyncer.tiangolo.com/tutorial/soonify-return/

**Comment by dubslow** [comment]
> Hmm.. I have to say I still don't quite get the problem. Like, this is what I think of when I think "asyncio.gather" to try to demonstrate what I mean by my previous pattern (I haven't tested this works, for what it's worth):

Essentially, yes I'm talking about the `gather` pattern such as you write, but with some more reflection upon this thread and the many links in it, I think the use of such `gather` recipes is hiding the fact that `trio` silently discards part of the control flow of the nursery -- the return values of task-function. That is, nurseries silently ignore half of the function interface, and considering that the function is a core block of Structural Programming, nevermind Structured Concurreny, it seems to me that `trio` violates its core principle in so doing.

In other words, I find the `gather` recipe to be only a hack covering a basic flaw in trio nurseries. Self-implementing the semantics is error-prone, not remotely thread-safe for example, and is exactly what a library is supposed to do: do the hard, common stuff right so the user doesn't have to. This hack of referring to a variable outside the nursery execution scope is exactly  what I referred to as an "extremely smelly ... considerable anti-pattern" in my previous comment. Implementing the task-function interface, including the collection of  return values, is the nursery's job, not the job of the scope containing the nursery, as in your example or mine or a dozen others that can be found on the internet. It looks far too much like the use of a global variable in oldschool threading, and carries exactly the same flaws, namely thread safety among others.

> I think I agree with @smurfix that there are so many specific things you might want (asyncio.gather (which has a boolean flag for operation), asyncio.as_completed, asyncio.wait (which overlaps with asyncio.gather with a 3-member enum flag for different types of operation)).

Being new to `async` in general, and having been persuaded by NJS's blogpost that I should therefore *start* with Structured Concurrency and ignore `asyncio` entirely, I'm not fully sure what those functions are.

That said, based on a quick Ctrl+F on the `asyncio` docs, they look considerably more complicated than what I say trio is missing. For instance, I believe that any talk of `Future`s automatically implies non-structured concurrency -- one of many goals of trio is to make such things redundant. Or for another example, `as_completed` looks like a crappy version of memory channels. Well this is certainly doing a good job of convincing me that NJS is right in the first place, and that trio is better than asyncio for writing concurrent code. (If by `outcome.Outcome` you mean [this](https://outcome.readthedocs.io/en/latest/api.html#outcome.Outcome), then I'd say that looks exactly like an `asyncio.Future`, a non-structured item, and whoever wrote it doesn't understand the meaning of structured concurrency. But then again, am noob, so take me with a grain of salt.)

All I'm saying is that nurseries silently discard half of the interface of task-functions, and discarding function interfaces is precisely non-structured, the very thing that trio is supposed to *not* be. Nurseries should store task-function results, and make them available to the containing scope after the nursery has exited (after the `with` block). Not doing so discards half the task-function interface, breaking the ideal of structured concurrency. The whole point of structured concurrency is that I should be able to easily trace control flow, including data that crosses stack boundaries -- arguments and retvals -- at all times by merely glancing at the code. 

Writing my own gather recipe by means of not-local-to-nursery variables is exactly the sort of anti-pattern that structured concurrency is meant to avoid. Re-phrasing it, we should be able to think of the `with open_nursery() as nursery:` block as its own function, its own scope so to speak, and thus all these gather recipes using closures, nonlocal references to stuff *outside* the nursery block, is exactly the sort of code that trio is supposed to prevent. And the nursery can, and should, store the return values for use *after* the `with` scope/block completes and exits, without use of "closures".

(I intended to not write a blogpost in a random github comment.... mission failed lol)

> I think, you may just use https://github.com/tiangolo/asyncer It is built with anyio, so it's trio-compatible
https://asyncer.tiangolo.com/tutorial/soonify-return/

That looks again like `asyncio.Future` or `outcome.Outcome`, the sort of thing that structured concurrency is supposed to prevent. If I need the results *inside* the nursery, I use memory channels, which is what channels were designed for. If I need the results *after* the nursery, then I shouldn't be writing that code inside the nursery with funny non-structured objects like Futures or Outcomes or SoonValues, which are of course all the same thing by different names.

**Comment by dubslow** [comment]
Let me try to summarize. Trio is meant to implement structured concurrency, where we can trace control flow (including stack/function call data) at any time, easily, by merely glancing at code.

Nurseries are the essential means of translating between synchronous and concurrent code, as per the neat little graphs in NJS's blogpost. The `with ... as nursery:` is the essential block: nurseries are in many ways akin to their own function call on the stack (despite not using Python's function call syntax), except it's a function call whose control flow arrows diverges into many such arrows in parallel, which nevertheless returns all arrows to a single point upon the exit of the `with ... as nursery` block. This is of course the entire purpose of structured concurrency, that all arrows have a start and an end -- and furthermore, nurseries have exactly one entry point and one exit point, no matter what the several arrows do inbetween.  The nursery manages the transition from syncrhonous to concurrent *and back* to synchronous. (The "and back" part is where most other async libraries fail to conform to structured concurrency, as NJS writes -- dangling arrows, silently discarded exceptions etc.)

The trouble is that, as it currently stands, nurseries discard task-function return values, which leaves each of its internal arrow *half*-dangling. When the nursery exits we know where the control flow is, so the arrow is partially complete, however some of the function-interface data, some of the stack data, is silently discarded, breaking the "functions as black boxes" primitive implied by "structured". In effect, each arrow of the nursery is only half-completed, the other half discarded and dangling. Upon the exit of the `with ... as nursery` block, the nursery's data is gone, those task-function retvals never to be recovered. Loosely speaking, I can trace the flow of the instruction pointer, but not the data pointer. I have no way to communicate my results from inside the concurrent nursery block back to its containing synchronous scope, after all the parallel arrows have reconverged onto the single exit point of the nursery. 

The use of a `gather`-like pattern, which are a dime a dozen on the web, with several examples here in this thread, is a hack by which one can fake their way into similar functionality. Nevertheless, it is not safe pattern, essentially by using a closure outside the `with ... as nursery` scope -- *a closure crossing between synchronous and concurrent contexts!!* -- to store those missing retvals for use after the `with ... as nursery` scope exits, after we reconverge from concurrent back to synchronous execution. Such closures break the "functions as black boxes" fundamental idea of structured concurreny, and is why they aren't thread safe, and generally considered bad practice, even in strictly-synchronous code. Closures break structure, and doubly so when such a closure crosses the synchronous-concurrent boundary as implemented by nurseries. In other words, much in the same way that "`goto` considered harmful" implies "`go` considered harmful", so too does "`go` considered harmful" imply "gather-like-recipes-doing-closures-across-nursery-scopes considered harmful".

All these gather recipes break the fundamental idea of structured concurrency. They harm the programmer's ability to abstract their code into components, which harms their ability to reason about their code and their ability to write it. Instead, since it's the nursery's job to manage the transition from one arrow to many arrows *and back* to one, it's also the nursery's job to pass data along both of those transitions. But, currently, the nursery passes data across one of those transitions, but not the other. (Specifically, `nursery.start_soon` accepts arguments which are forwarded to task-functions: here, the nursery does its job of passing data from the one single entry point into the many parallel arrows, but the nursery later discarding the retvals of the same functions means that no data ever passes from the parallel arrows back to the single exit point, breaking structured programming.)

So, in short, nurseries are not yet complete implementations of structured concurrency: the fact that all these `gather`-like recipes use closures (at least in their shortest form they do), not to mention the fact that *those closures cross synchronous-concurrent boundaries* (!!), is a big hint that these `gather`-like recipes are in fact covering for a big flaw in nursery's implementation of structured concurrency. (Of course, one could simply pass the synchronous results collector into the nursery and its tasks, which would make it technically not a closure, but concurrent tasks all writing to a synchronous variable remains a seriously smelly anti-pattern, breaking the abstraction principles of structured concurrency, no matter if it's technically a closure or not.)

**Comment by TeamSpen210** [comment]
In regards to creating a function for each task, that works and is a quite powerful approach, but it does have the problem that these functions are inherently specialised and can't be reused. It also requires a full block syntax, so if you have a bunch of heterogeneous tasks it gets a bit awkward to make these every time. 

`outcome.Outcome` is somewhat different to the other objects you've mentioned actually, because it's not based in concurrency. It's simply a container to hold either an exception result or return value, then unwrap that elsewhere. What I've been using personally in my own code is a similar sort of container to that as well as `asyncer.soonify`. The difference is that it won't let you access the result until the nursery is completed successfully, so the result must always be available.

**Comment by dubslow** [comment]
> In regards to creating a function for each task, that works and is a quite powerful approach, but it does have the problem that these functions are inherently specialised and can't be reused. It also requires a full block syntax, so if you have a bunch of heterogeneous tasks it gets a bit awkward to make these every time.

I don't follow your meaning. Tasks *are* functions. Writing a task = writing a function.

**Comment by TeamSpen210** [comment]
The way outcome works is that you use the `unwrap()` method, which either returns a value when successful , or re-raises the exception if it failed. Trio uses them internally for tracking results a bunch. For [my helper](https://github.com/BEEmod/BEE2.4/blob/f2e4a629c4373ba283473fdbcbe469232fdf2de7/src/utils.py#L473), it's just used like so:
```py
async with trio.open_nursery() as nursery:
    # internally, nursery.start_soon(some_task, arg1, arg2)
    func_result = Result(nursery, some_task, arg1, arg2)
print(func_result() + 5)
```
It's also got a `sync()` classmethod to delegate to `trio.to_thread.run_sync`, since I use that a lot. Currently it's abusing the `CancelScope` `repr()` to be able to check if the nursery has completed successfully, I should open an issue for exposing that properly.

**Comment by dubslow** [comment]
Ah, very nice. That is indeed approximately the functionality that a nursery should implement by default. Ultimately I think it shouldn't be necessary to store some new variable from within the nursery, since what you have looks pretty close to a `Future` to me, but this is indeed fairly clean, at least from a user perspective.

Probably the smelliest thing about it is indeed the abusive way to track the nursery status. My "cleanest" hack stores the results on the nursery itself, which obviates that need, and removes the "storing a new variable inside the nursery bit", but then my hack still looks like several threads smashing a single global, which your faux-Future doesn't look like, instead leaving collation/serialization until after nursery exit.

Still would be preferable, and cleanest, to have this sort of semantics in nurseries from the start of course. We shouldn't have to add our own code, no matter the form, in order to get fully structured concurrency.

**Comment by smurfix** [comment]
> I don't follow your meaning. Tasks are functions. Writing a task = writing a function.

A task is a function with some arguments that you start at a specific time. One function = any number of tasks (including zero). Thus, using the function as an index for the task's result doesn't always work.

> what you have looks pretty close to a Future to me

Well, no. A `Future` is created, then fed with a result/exception some time later, then you wait on it, then you access the result or exception. An `Outcome` conflates the first two steps and cannot be waited on.

Yes a nursery could save the results of all those tasks. But how do you access them later? Sequentially? (By task start or by task stop?) By some key? Where and based on what do you create that key? Or simply by returning a `Future` object from `start_soon`? Is an exception a "result", and does it cancel the nursery, as it does now, or not? Do you need to be able to cancel the task? Is the task auto-cancelled when the object that's going to capture its result is no longer accessible, and if so what happens if it raises someting that's not a `Cancelled` exception?

The answer to these questions is highly application specific, as is the "OK I have five nurseries and exactly one of these needs to go through this dance while two others absolutely must not" problem.

It's rather easy to write a nursery wrapper that answers each of these questions in one specific way. Writing one that does all of the above (and believe me I've needed all of them during the last few years) might be possible, but then you need to decide each time which variant you want (including the don't-save-a-result-dammit version). Far easiler, IMHO, to have a nursery that drops results on the floor, and write a local, small and *obvious* helper that saves your results to wherever when you need that.

**Comment by dubslow** [comment]
> Yes a nursery could save the results of all those tasks. But how do you access them later? Sequentially? (By task start or by task stop?) By some key? Where and based on what do you create that key?

By default, a nursery should do the simplest thing, which is to simply sequentially store for them for access after the nursery exits. Anything fancier can be done by the end user encoding their retvals as they need. In other words, you're overthinking it. The nursery just needs to provide any retval at all, and the user can build atop that with no problem. Currently, there's nothing to build atop, which is the issue: violation of structural programming.

> Or simply by returning a Future object from start_soon?

The whole point of this thread is that the return values are meaningless until the nursery has exited (reconverged from concurrent to synchronous). A format like this gives the wrong idea.

> Is an exception a "result", and does it cancel the nursery, as it does now, or not? Do you need to be able to cancel the task? Is the task auto-cancelled when the object that's going to capture its result is no longer accessible, and if so what happens if it raises someting that's not a Cancelled exception?

Well frankly this these are orthogonal questions. As it is, users have no way of (easily) customizing cancellation of a nursery, so these questions are kind of moot. Ultimately, I don't see anything wrong with the simplest answer: the nursery stores any retvals it gets, and those are available to the user after the nursery has exited. Any tasks which were interrupted, which never returned, are therefore not in the results, and it's up to the user to handle exceptions, exactly as currently. The answer to these questions is completely independent from the nursery storing task retvals.

> The answer to these questions is highly application specific, as is the "OK I have five nurseries and exactly one of these needs to go through this dance while two others absolutely must not" problem.

Nurseries are little customizable as-is, so your concern is already a concern, quite independent of the question of nurseries breaking structured data flow or not.

> It's rather easy to write a nursery wrapper that answers each of these questions in one specific way.

"nursery wrapper" is exactly the sort of phrase that should clue you that something is wrong. Wrappers crossing the synchronous-concurrent boundary, i.e. crossing into and out of the `with ... nursery:` scope, are exactly the sort of anti-pattern which breaks structural programming. All I'm asking is that nurseries respect structural programming, as they're supposed to. Nothing else changes, it's no earth-shattering revelation.

> Far easier, IMHO, to have a nursery that drops results on the floor, and write a local, small and obvious helper that saves your results to wherever when you need that.

There is *nothing* that is remotely *obvious* about any of the hacks listed in this thread. The whole point of this thread is that **they are hacks, harmful anti-patterns which degrade programmer ability to abstract and reason about the code**. Dropping stuff on the floor is the worst thing a nursery can do. Seriously, Structured Programming 101 right here.

**Comment by smurfix** [comment]
> By default, a nursery should do the simplest thing, which is to simply sequentially store for them for access after the nursery exits. 

Which sequence – task start or task exit? What about nurseries that run forever, do they contain a possibly-infinite array of `None` values?

Why build that into a nursery when you can add three lines of code to wherever you actually need it?```
res = []
async def save(p,*a):  # store task results, task-exit order
    res.append(await p(*a))
async with open_nursery() as n:
    n.start_soon(save, fn, args…)
```

> The whole point of this thread is that the return values are meaningless until the nursery has exited (reconverged from concurrent to synchronous). A format like this gives the wrong idea.

Well, my point is that sometimes you need the values as they are generated, sometimes you don't need the overhead of saving them, and sometimes saving them is actually harmful.

"default to saving results" thus causes your code to work quite happily when you test it, but to eat an unbounded amount of memory in production unless you add an explicit option telling it not to. Talking about anti-patterns, that's a fairly major one for me.

> Well frankly this these are orthogonal questions

If you want an exception to be saved in an `Outcome` object for later perusal you need a function just like the aforementioned `save`, except that it calls `outcome.acapture(p,*a)`. Ditto (more or less) for answering the other questions in that block. 

Thus, as soon as you get around to implementing any of this, it's hardly orthogonal.

> "nursery wrapper" is exactly the sort of phrase that should clue you that something is wrong.

Might I suggest that you re-evaluate what I wrote?

A nursery wrapper for me is simply a class that contains a nursery (as opposed to a nursery subclass, which Trio doesn't allow – for reasons I don't entirely agree with, but that's orthogonal to this topic) which passes everything through to the "real" nursery, except that it augments `start_soon` and `start` in order to add the aforementioned `save` method/function. If this method stores the result in a `nursery_wrapper.results` attribute you get exactly what you're advocating for.

I for one am perfectly happy to add that simple `save, ` to those `start_soon` or `start` calls that actually need them. Explicit is better than implicit, simple is better than complex, and all that. IMHO.

**Comment by dubslow** [comment]
> Which sequence – task start or task exit?

Considering the former is nonsensical, the latter.

> What about nurseries that run forever, do they contain a possibly-infinite array of None values?

That's up to the user to not launch an infinite number of tasks.

> Why build that into a nursery when you can add three lines of code to wherever you actually need it?```

As I've repeatedly stated (but allow me to repeat again), this "pattern", which is in fact an anti-pattern, is extremely harmful. Your `res` variable crosses the synchronous-concurrent boundary which is the whole purpose of nurseries to enforce. In other words, this idea breaks the whole idea of nurseries. It ruins programmer ability to abstract and reason about the code, leading to bugs. Not surprisingly, it's also literally dangerous in the sense of not being threadsafe, being equivalent to many threads accessing a global variable at the same time.

> Well, my point is that sometimes you need the values as they are generated, sometimes you don't need the overhead of saving them, and sometimes saving them is actually harmful.

> "default to saving results" thus causes your code to work quite happily when you test it, but to eat an unbounded amount of memory in production unless you add an explicit option telling it not to. Talking about anti-patterns, that's a fairly major one for me.

The *entire purpose* of a nursery is to enable many ~independent tasks in parallel. The user can spawn as many tasks as they like to consume as much memory as they like. Nurseries fully implementing structured programming doesn't change the fact that they *already* allow the user to do bad memory management. What it *does* change is the ability to avoid harmful global-like patterns such as the one you noted. Your other concerns already exist as-is.

> If you want an exception to be saved in an `Outcome` object

Huh? Since when are exceptions special? Exceptions cancel the nursery currently, and that wouldn't change.

> Might I suggest that you re-evaluate what I wrote? A nursery wrapper for me is simply a class that contains a nursery

Indeed, which is to say, it's not much different from your `save`  and `res` pattern, abstractly speaking. It breaks the synch-concurrent boundary by "containing" a nursery and implementing new semantics both inside and outside that nursery. And by crossing that boundary becomes dangerous, harmful, breaking structured programming.

> I for one am perfectly happy to add that simple `save, ` to those `start_soon` or `start` calls that actually need them. Explicit is better than implicit, simple is better than complex

**This `save` pattern is not simple, but is in fact copmlex, dangerous and harmful.** The *appearance* of simplicity is part of its harmfulness. It is not simple at all! It's like a global variable, and carries all the dangers of global variables in a threaded context. Don't let the fact that "async event loops are single-threaded" confuse you into believing that this `save` pattern isn't dangerous. It mixes concurrent and synchronous code in the worst possible way, in exactly the way that trio tries to prevent. This `save` pattern violates everything that trio is supposed to stand for. And by enabling users like you to think that this is okay, or even good, is why this gap in nurseries should be fixed ASAP, IMO. Fake-gathers considered harmful.

**Comment by smurfix** [comment]
>>    Which sequence – task start or task exit?

> Considering that the former is nonsensical (saving a value that doesn't exist yet), the latter.

What's nonsensical about that? When I start task A and then task B, I know which order I started them in but not which order they finished. So I might want to access task A's result by accessing `nursery.result[0]`.

> Your res variable crosses the synchronous-concurrent boundary which is the whole purpose of nurseries to enforce.

Well, that's your opinion. Mine is that while control flow and data flow are of course causally related, they're otherwise independent concepts, and I see no good reason why a nursery's job should be to **enforce** any aspect of the latter.

**Comment by dubslow** [comment]
Well "enforce" is always a strong word, Python users can easily shoot themselves in the foot with any code if they try hard enough, but in this case I just think that nurseries should give dataflow the same support as controlflow. It should be possible for users to easily obey the rules (even if Python doesn't really "enforce" rules properly).

But I appreciate that we can politely agree to disagree :+1:

**Comment by tapple** [comment]
I kind of disagree with the entire premise of this issue: that nurseries are like a function call. I think they are not. I think they are more comparable to an if or for statement. an if does not have a return value. nor does a for loop. Nurseries are not a language primitive at this time, so they are built out of something that is: functions. However, the functions that make up a nursery are an implementation detail, and nurseries are under no obligation to behave like functions and have a return value

**Comment by A5rocks** [comment]
A few days ago I was writing code and had to get the return values of these two functions I was running concurrently. I ended up just using some globals since it was easiest, which isn't great! (this isn't exactly a gather typing-wise though)

I still disagree with the fundamental premise of "nurseries need to `.join()` return values too", but I think it would be *really convenient* if we had something along the lines of what @TeamSpen210 mentioned. Here's what I had in mind from before reading through this again:

```py
async def f():
  await trio.sleep(2)
  return 5

async with open_nursery() as nursery:
  f_result = nursery.start_soon(f)
  f_result.unwrap()  # would error, as `f` is still running
  await trio.sleep(5)
  f_result.unwrap()  # would error, as the nursery is still running

print(f_result.unwrap())  # prints 5
```

We could make this essentially no extra overhead using some weakrefs (`Task` -> this `TaskResult`, `TaskResult` -> `Nursery` to check if nursery is running (?)). I think this avoids many of the possible questions mentioned above since you can still create some different gathers with it.

Here's some upsides of this over some of the designs here:
 - minimal overhead unless you want it (you can still have infinite tasks in a nursery!)
 - you can design `gather` sorted on task start time / task arguments. (though see downsides for sorted on task end time). this isn't much of an upside since the implementation without this is quite simple.
 - typing becomes trivial
 - prevents some race conditions due to not working while the nursery is running

Here's a list of things to consider:
 - `nursery.start()` already returns a value. We would have to have to add a kwarg there which changes the return type
 - naming it (`TaskResult` isn't a great name, `Result` is taken by `outcome`, this is not a `Future`)
 - what should it return if `f` ends with a `Cancelled` exception? every other exception is obvious: `.unwrap()` should raise it. I guess it should also raise `Cancelled`?
 - implementing some types of `gather` isn't easy: if you need task end time... well it's not possible without a wrapper at which point you could just `.append()` a list.
 - I had some other things I think but they slipped my mind while writing this up.

@smurfix does this seem like a convenience feature that would help in several situations? Even though it's minimal overhead, it's extra API surface area so I think this should still be applicable in many places.

**Comment by GalaxySnail** [comment]
What about [aioresult](https://github.com/arthur-tacca/aioresult)?

**Comment by TeamSpen210** [comment]
For `Cancelled`, I reckon it should raise an error - the idea of this is that you're only allowed to call `unwrap()` it's certain the task has finished successfully, so directly cancelling the nursery its in and expecting to extract a result isn't really right. Perhaps the cancellation itself would be an error?

**Comment by A5rocks** [comment]
> What about [aioresult](https://github.com/arthur-tacca/aioresult)?

That's more of a future (`ResultCapture` has `wait_done`) and I have a feeling `ResultCapture#result()` being usable *during* the nursery is something better not exposed. Otherwise it spells a few other things differently but `ResultCapture` is close to what I describe.

> Perhaps the cancellation itself would be an error?

That feels like a bit much. I guess throwing `Cancellation` makes sense.

**Comment by GalaxySnail** [comment]
> * what should it return if `f` ends with a `Cancelled` exception? every other exception is obvious: `.unwrap()` should raise it. I guess it should also raise `Cancelled`?

IMO `.unwrap()` shouldn't raise a `Cancelled` exception, because `Cancelled` shouldn't escape out of its cancel scope. Since cancellations are level-triggered, it shouldn't be a big problem if it's cancelled from outside. Instead, if a task is cancelled, it's better to raise a new `TaskCancelledError` in `.unwrap()` to indicate that it has been cancelled.

**Comment by TeamSpen210** [comment]
Oh, just realised something from your description. I don't think this `TaskResult` object should store exceptions at all. If the task raises anything, it's simply propagated as normal, which would ideally prevent code reaching the `unwrap()` call.  Only time you'd get something meaningful is if the nursery completed successfully, meaning the task did too.

**Comment by A5rocks** [comment]
I mean it's still possible to get to a relevant place and it's probably worth it to have some defined behavior there (and it would be very nice to have the invariant of "if you have a `TaskResult`, you can get how the task exited" (regardless of success or failure or cancellation or ...):
```py
try:
  async with open_nursery() as nursery:
    result = nursery.start_soon(blah)
except:
  print(result.unwrap())
```



> Instead, if a task is cancelled, it's better to raise a new `TaskCancelledError` in `.unwrap()` to indicate that it has been cancelled.

Yeah I was thinking that but... at that point why not just raise `Cancelled`? To try to allow someone to catch it with an `except Exception`?

**Comment by TeamSpen210** [comment]
I guess the problem is that having the result store the exception implies that it captured that exception - but it didn't, the exception was already raised. And as we've seen elsewhere re-raising an exception stored like this is liable to wreck the traceback, which this would encourage doing. It just feels inconsistent with the rest of nurseries - other than cancellations, nurseries never save or catch exceptions, just merge multiple streams together. Similarly, there's no way to say have a comprehension return like a partial list or something.

Actually, if you do want to catch exceptions, that's easy to do, just pass `Outcome.acapture` to the nursery. Would that be a good solution? It's not something I imagine needing often...

**Comment by smurfix** [comment]
> Yeah I was thinking that but... at that point why not just raise `Cancelled`?

Because at that point the cancel scope that originally caught the cancellation responsible for the error has already ended. Thus Trio's standard `Cancelled` exception would not be appropriate. You should raise some other exception at this point, preferably one that's descended from `Exception`, not just `BaseException`.

**Comment by TeamSpen210** [comment]
Raising cancelled is also sorta problematic because any cancel scope could potentially swallow it. Say for example you opened a nursery, made a result, cancelled the nursery, but then did something else which caused the surrounding task to also be cancelled. Then `unpack()` would suddenly act like a checkpoint and cancel out this task. Could be tricky to debug.

**Comment by A5rocks** [comment]
Makes sense to just ignore exceptions if this is the level of complexity we would have to deal with. (Honestly just the point about tracebacks mutating is enough to scare me off). I guess the main thing I'm unsure about then is:

```py
async with open_nursery() as nursery:
  res1 = nursery.start_soon(trio.sleep, 1)
  res2 = nursery.start_soon(trio.sleep, 5)
  res1.unwrap()  # should obviously raise UnfinishedTaskError
  await trio.sleep(2)
  res1.unwrap()  # should obviously raise UnfinishedTaskError
  nursery.cancel_scope.cancel()

res2.unwrap()  # should obviously raise UnfinishedTaskError
res1.unwrap()  # ... should this raise?
```

I'm partial to not raising because that sounds complicated to implement but it feels less confusing if it were to also raise? Am I just talking myself into a corner?

(also `.unwrap()` should probably be a different name. I don't like a weakref style anonymous call though)

**Comment by smurfix** [comment]
> I'm partial to not raising

So am I, for the (IMHO very good) reason that the subtask in question has terminated with a result and thus passed the responsibility of dealing with the result to the task that invoked it.

It's counterintuitive to not have that value available, not to say very annoying when you need it anyway and now have to fight Trio for it.

>   await trio.sleep(2)
>   res1.unwrap()  # should obviously raise UnfinishedTaskError

That's not so obvious to me. Task 1 has returned a value. Maybe it has even set an event that tells the main task that it did so. Now why should the result be withheld until the nursery has terminated?

**Comment by A5rocks** [comment]
Basically to try to avoid some basic race conditions, as well as strongly indicating that these aren't futures:tm:. Maybe that's not a great argument for, but I really want to avoid implementing some sort of future in Trio.

**Comment by TeamSpen210** [comment]
Perhaps it should be a `.result` property. Result ensures we could put other values later, and it's a property because it's cheap to fetch, no synchronisation/waiting required. I was going to say it might make sense to have `res1` not raise, since a common concurrency pattern is to have racing tasks, then cancel after the first wins. But it wouldn't actually be useful there - first you need the tasks to tell you which succeeded, so they might as well also pass out the result directly. And you need to also handle the case of multiple tasks succeeding simultaneously. Not really sure if there's a case where you know the task has a result despite the cancellation "statically". Since my conception of this is that the lexical location of the access proves the task has a result race-free.

Sorta simpler to just say that to use it the nursery must have completed successfully, and a cancelled nursery didn't succeed. I'm in favour of raising just one exception for all failure modes, with the message indicating it (since you shouldn't be catching it). So in your example, inside the nursery it'd raise `NoResultError: Nursery has not exited`, then afterwards just `NoResultError: Nursery cancelled`.

Just had another idea though. For the situation where you do want values early, what seems very trionic is to make this result object have a `.out_to(send_channel)` method. That'd effectively graft `await send_channel.send(retval)` onto the task. You could then just do `nursery.start_soon(some_func, args).out_to(some_channel)`. Or I guess it could be a keyword argument on the nursery.

**Comment by richardsheridan** [comment]
>Just had another idea though. For the situation where you do want values early, what seems very trionic is to make this result object also have a .out_to(send_channel) method. That'd effectively graft await send_channel.send(retval) onto the task. You could then just do nursery.start_soon(some_func, args).out_to(some_channel). Or I guess it could be a keyword argument on the nursery. This would be in addition to the property.

I'm for this kind of thing being a only feature of the nursery itself rather than the `start*()` methods or some retval type.  For example, `trio.open_nursery(result_channel=send_chan)`.

This strategy reminds me of one of our other APIs, `start_thread_soon`. Here, NJS noticed that it's really handy to get the result of a thread even though it is typically punted into the nether by the Python `threading` API. So, as a sync API, a callback made sense to get the result asynchronously and help get it back into the Trio event loop. 

In the present case, we have an API in the `Nursery` to deal with not just the result of one task but several tasks. We just shouldn't give the nursery a callback kwarg in the constructor, since that's against structured concurrency. But a channel will do the trick instead! The nursery already waits for the task to finish, including gathering exceptions, so it doesn't seem like a stretch to have the nursery be the thing that does `await send(result)` as well. 

This way, the nursery doesn't need to be decorated with extra attributes or methods (or methods on methods?!), and we don't need to invent a new not-quite-future type... just use the existing channel infrastructure! You want to build up a list of the results like `gather`? pass a inf buffered channel! You want to cause arbitrary async work depending on what's coming out of the tasks? Put the other end of the channel in some external task!

Grafting `await send_channel.send(retval)` onto the task hasn't been that hard so far, but this would maybe help nudge people in that direction?

**Comment by TeamSpen210** [comment]
Not quite, because personally the main use case for this feature is for heterogenous nurseries - for example, start a bunch of tasks to read in different config files simultaneously, and then once they're all done do stuff with each result. This really should be per-task. `nursery.start_soon(func, args, result_to=some_channel)` would possibly work there, though it'll be awkward - you'll need 1 line to make the channels, 1 to start the task, and if you want to use the value multiple times you'll need a 3rd to store it. Plus 3 different variables to keep track of.

That's why I'm thinking there should be both options, for different situations. The result property is sync and cheap, because it can rely on the nursery to ensure the task is done first.

**Comment by A5rocks** [comment]
Yeah to be clear about the framing I have (i.e. the place where I encountered this), I was building a thing to scrape a site to download information to a local DB. I wanted to scrape two different classes of things in parallel, as well as fetching follow-up links from one of those things (with a bunch of tasks at once). Most of that was easy, but to actually get the fetched information back, I ended up appending to a global array for convenience sake.

So I think it's most convenient (which is ultimately what this is about, it's not like we're unlocking any new capabilities) to have the information generally segregated, instead of shoving it all into a single channel and seperating it out later. Also having it per-task lets us nullify most overhead from storing unwanted results.

---

Memory channels are an interesting idea. I really don't like the feel of using them there, but I can't think of concrete reasons why not. I think I just really am put off by anything that notifies the parent task that their child task has finished executing, for some reason. I can try to rationalize it as: if you are building a complex system that has links between child tasks/between child tasks and the parent then you probably don't care about the 5 lines extra it currently takes to do the same... but I know that's not actually why I don't like it.

**Comment by smurfix** [comment]
> once they're all done do stuff with each result

You could do that with a nested nursery and/or a channel that processes each result as it arrives (and which auto-closes after all producers have terminated, which is almost comically easy to achieve with memory channels).
I'd argue that this makes more sense than thinking in terms of `asyncio.gather`.

The problems start when you have *one* task whose result needs to be processed somehow. Maybe A fetches some data which B preprocessed and forwards to C, but the setup code for B+C is sufficiently complex that I want to parallelize most of this.

The best way to do that seems to be a `nursery.start_task` method that returns an outcome+event object, thus kindof like a Future, but bound to the nursery so that errors cannot get lost. We could even bind a cancel scope to it …

> I really don't like the feel of using them there

Neither do I to be honest. But IMHO the main questions we should ask is "does this violate SC" and "does it involve callbacks". If the answers to those questions are "no" we should (again IMHO) strive to implement the most generic solution possible, i.e. the one that affords the most usage patterns. Results which aren't available during the nursery's lifetime, much less ones you can't wait for, don't cut it. They just lead people to reinvent unbound futures and similar buggy anti-patterns. My own codebase contains plenty of those things and I'd like to get rid of them.

Here's a quick sketch of how that would look like:

https://gist.github.com/smurfix/9c986d3f20beb5e55b9c9404e795e5ed

**Comment by TeamSpen210** [comment]
That solves an entirely different problem, and really wouldn't work for the sorts of things I'd want for example. Your future catches exceptions, isolating all the tasks. I don't think a nursery should be doing that, the only things it catches are cancellation exceptions (which aren't really errors or anything). If you want to be catching exceptions, that can be done easily by wrapping your task in `acapture`, then using this result mechanism to retrieve the `Outcome`. The other way around would require adding dummy `result` calls just to raise, which also would cause exceptions to be raised in a different place to where the function was actually called.

To compare, here's a [good example](https://github.com/BEEmod/BEE2.4/blob/bd62a0d12fca832ddf4e1f708f27c38dc6ad04aa/src/vbsp.py#L1634-L1638) of where I'm using `aioresult.ResultCapture` for this purpose. I spawn a bunch of tasks which each parse a different config file, some directly . These are all heterogenous types, and obviously I don't want them mixed with each other. Additionally if any fail to parse, everything should fail.

**Comment by smurfix** [comment]
> Your future catches exceptions, isolating all the tasks.

Well sometimes that's the point. Sometimes I do want to handle an exception at the place where the result needs to be processed. The point of this exercise is to not let an unfinished future-or-however-to-call-it *or* an unhandled error escape the nursery.

Teaching my code to optionally not prevent the nursery from getting cancelled when there's an exception is easy enough, and so is to not catch errors at all. Gist updated.

**Comment by arthur-tacca** [comment]
Disclosure: I'm the author of aioresult, so naturally I agree with all its design decisions!

Is there really anything needed that isn't already solved in aioresult? It already exists.

```python
async with open_nursery() as nursery:
  f_result = ResultCapture.start_soon(nursery, f)
  f_result.result()  # would error, as `f` is still running
  await trio.sleep(5)  # or - await f_result.wait_done()
  f_result.result()  # would NOT error, unlike the comment above

print(f_result.result())  # prints 5
```

* There was an objection that `ResultCapture` has `wait_done()`. But I don't see why that's a problem if you don't need it. It doesn't even instantiate an `Event` object unless you call `wait_done()`. It just allows some extra valid use cases (like the next bullet point). I can see an issue with users needlessly waiting if they don't need to if you have a single `wait_and_get_result()` (like smurfix's recent example) but I deliberately return `None` from `wait_done()` to avoid that mistake.
* There was an objection that `ResultCapture.result()` works even if the nursery isn't done yet. Again, I don't see the problem. It unlocks some interesting use cases, where you run multiple tasks in parallel, which are consumed from other tasks, etc., all in one nursery.
* There was discussion about whether the result object should re-raise the original exception, which may not make sense in the new context, especially for `trio.Cancelled`. `ResultCapture.result()` deliberately doesn't do that so you don't have that problem. Instead, it raises [`aioresult.TaskFailedException`](https://aioresult.readthedocs.io/en/stable/result_capture.html#aioresult.TaskFailedException). The original exception is accessible as the `ex.__cause__` property (and `ex.args[0]` has the `ResultCapture` object). If you don't want to raise anything at all when there was an exception then you can check for it manually (`rc.result() if rc.exception() is None else None`).
* As a compromise, `ResultCapture` optionally lets you specify that exceptions should not be propagated to the enclosing nursery ([`suppress_exception=True`](https://aioresult.readthedocs.io/en/stable/result_capture.html#aioresult.ResultCapture)) but even then that still passes up exceptions directly derived from `BaseException` (e.g. `Cancelled`).
* There was brief discussion about whether `rc.result` should be a property. As it was in aioresult 0.9! I changed it to a method because it can raise `TaskFailedException` or `TaskNotDoneException`.
* There was a mention of saving task results directly in the nursery. "That's up to the user to not launch an infinite number of tasks." No, launching an unbounded number of tasks (over time, not all at once) is a valid use case for nurseries.
* Per-task cancellation: I like this idea but didn't initially add it to aioresult because njsmith suggested it would be a bad idea in general (something like "cancellation scopes are cheap but not free"). It would be easy to add as a subclass but I just haven't got round to it ([issue #2](https://github.com/arthur-tacca/aioresult/issues/2)).
* There was a suggestion of using a stream to communicate the results: 
  * This could be useful sometimes but annoying in simple cases where you just want to save a couple of results in local variables, and doesn't generalise to complex situations where multiple things waiting for one task (they can't both iterate over the steam looking for what they want).
  * I have a little utility function in aioresult, [`results_to_channel()`](https://aioresult.readthedocs.io/en/stable/wait.html), that lets you put results into a memory channel, though it only works with a fixed number (not dynamically adding more as the nursery runs).
  * I did also make a variant that allows dynamically adding tasks and considered adding it to aioresult; it's [in this gist](https://gist.github.com/arthur-tacca/5c717ae68ac037e72ae45fd1e9ca1345). But I soon found it's surprisingly subtle to decide when to close the stream when nurseries interact in a complex way. I don't think there's a perfect solution for all cases.
* I think the number 1 weakness in aioresult is that it's not just a simple return value from `Nursery.start_soon()`. But we come back to the problem of what to do about `Nursery.start()`. njsmith [mentioned in a comment](https://github.com/python-trio/trio/issues/410#issuecomment-359612880) that `start_soon()` actually did used to return a task-like object for fetching results, and the `Nursery.start()` API issue was a big reason for him getting rid of it. 

----

Sorry for the huge wall of text, but I've tried to understand every point. I'm not precious about the design of aioresult but, so far, I haven't seen any convincing arguments for doing anything differently. Is there anyone that really couldn't do what they need to by just using `ResultCapture` as it is?

It seems the biggest problem is just awareness. If there were consensus that aioresult is the right approach (which I feel like, at this point, there is not) then we could just add a note about it to the Trio task spawning docs. It's already mentioned in Awesome Trio Libraries. I don't think it would be right to incorporate directly into Trio (though I'm not dead against it) because we've seen it's so hard to get agreement.

**Comment by A5rocks** [comment]
Had I known about aioresult and its specific design choices earlier, I would have absolutely landed to where I currently am faster. But we're here anyways so what's the harm! Point by point:

- At least on my part this is a knee-jerk reaction to futures and trying to evade having them in Trio. I think what a library exposes is important and if a pattern is 50% right or 75% right but encourages some bad habits, then it probably shouldn't be here. I'll try to write up a stronger rationalization at some point (or in doing so, talk myself out of my position).
- Similar as above; tying whether you can access a result to a lexical scope means avoiding some race conditions is feasible. Obviously you can sidestep this by sending a result to a child task started on a parent nursery, at which point you opt into all the race conditions. But that's detectible using static analysis unlike "am I calling `.result()` after my task of arbitrary length is over" (in which case you may want an async `wait_and_get_result` to avoid things, at the cost of having more async)
- Setting `__cause__` is good but the traceback might be misleading.
- Opt-in via composing `outcome.acapture` probably makes more sense for above traceback reasons. (even though we could do some magic to avoid it)
- This is good!
- Absolutely.
- Yeah this makes sense too, extra frames is Bad. I'd love to have concrete numbers to work with.
- comments on vs stream
  - Yeah I agree that they are worse in "just give me this value!" and IMO that's what we should focus on.
  - No comment.
  - Thanks for the implementation information! I somewhat ran into this when I was thinking out updating happy eyeballs (for which streams, and not above results, seem uniquely useful for). I couldn't figure out where "await the result" would sit.
- I really need to get around to reading all the old issues instead of my current piecemeal approach, thanks for the link. I somehow didn't see that even though I did a basic search in the issue tracker for "future" (there were a lot of unrelated results...). I asked above what to do about `nursery.start` and nobody added extra ideas to the one obvious "just add a kwarg" idea.

As seen above, I think I generally agree with the design decisions you made, except for anything where I just don't feel right about it.

**Comment by TeamSpen210** [comment]
For `start()` I'm not sure it's too necessary. One main advantage of having this is that you can use it with any async function, you don't need to define something bespoke each time for nonlocals or patch in channel logic. But for a startable task that needs alteration anyway, and such a thing is usually long-running also. Pretty much all the ones I have end with an infinite loop of some kind. If you need to return things you have `task_status` right there.

I do think it probably should avoid letting you get results early, but results from successful tasks after a failed nursery isn't as bad. Cancelled ones especially. But it probably shouldn't store the exception I think.

**Comment by Sxderp** [comment]
Do we need to maintain the `nursery.start_soon` interface? Can the "gather results" be on a separate function call? That way you can store the results in/on the nursery without worrying about "saving everything" and having unbound memory usage.

I like the idea that results are only accessible after the nursery has closed. Isn't that the point of "gather"? It waits for everything and then gives a result. I tend to see people re-using nurseries when they should be creating new child nurseries and this can help re-enforce that.

I also feel the idea that you should be able to "await the results object" (within the nursery) doesn't make sense. Why not just await the function call instead of wrapping it..?

```python
import trio

async def sometask(): ...

async with trio.open_nursery() as nursery:
    nursery.gathered_start_soon('uniq_id_1', sometask)
    nursery.gathered_start_soon('uniq_id_2', sometask)
    nursery.gathered_start_soon('uniq_id_3', sometask)

    # it's an error to reuse an id...
    nursery.gathered_start_soon('uniq_id_1', sometask) # GatherIdReuseError

    # Error to try and grab something when nursery has not exited
    nursery.gathered('uniq_id_3') # NurseryNotExitedError

theresults = nursery.gathered('uniq_id_1')
```

I'm not sure how cancellation should work. Maybe a NoResultsError when trying to gather if the task was cancelled / unable to return?

**Comment by laclouis5** [comment]
> In many cases you want partial results from subtasks as soon as they're available. (Possible solution: Use a memory channel.) 

This is precisely the **opposite** of what `gather()` is for. `gather` is supposed to yield when *all* results are available, i.e. no partial results.

> Sometimes you also need to associate task names/objects with the results (send a name+result tuple to the channel.) 

This one can be addressed trivially by a `gather()` function since outputs are in the same order as the inputs. With `gather()` you don't even *have* to do this association.

> In others you need a list (write a two-liner that appends the result to an array) or associated with an object (store the result in the object, or write another two-liner that saves it to a dict.) 

This one is likely the default behavior of a `gather()` function: you get a list as output. Association with a corresponding object or to a key in a dict would also be trivial.

> In many others you don't need the return values at all and don't want to save a possibly-unbounded list of `None` values. And sometimes you need to combine any of the above with a concurrency limiter, which also might have to decide what to do about the return value.

It is perfectly possible to use `gather()` with functions returning nothing (`None`), and it is not even an anti-pattern IMO. It is designed to process a *finite* amount of coroutines, whatever they return. If you have a "possibly-unbounded" list of coroutines returning *anything* (including `None`) then `gather()` is definitely not for you.

**Comment by TeamSpen210** [comment]
@laclouis5 This discussion isn't solely about `gather()`, it's about the general problem of how to get return values out of tasks, where the input/output list of `gather()` is just one specific method which indeed does not work for many cases. Maybe I should have worded my link better.

**Comment by laclouis5** [comment]
@TeamSpen210 The fact that "[some method] is just one specific method which indeed does not work for many cases" is irrelevant since this is literally true for any function in any library. I find quite unfortunate that is argument is used over and over again in this thread and related others (#2188), especially for the `gather()`function. It's actually a good design pattern for a method to be "specific" and not pretend to cover every conceivable use-case.

I argue that `gather()` would cover *many* cases that are currently hard to address with Trio and thus would be a great addition. Surely it won't solve all use-cases and additional functionality such as a mechanism to get the result of tasks scheduled by a nursery would also be relevant.

**Comment by m-u-xyz** [comment]
The problem is, "covers many usecases" is not a good argument by itself. "goto" covered all the usecases of "while" and "foreach" and whatnot, yet we all decided that it's a very bad idea anyway – or rather, exactly *because* it did.

Yes `gather`is just one specific method and *of course* it doesn't cover all usecases. Yes it covers a whole lot of them. These are not mutually exclusive.

The question is not whether `gather` should be a first-class taskgroup/nursery method or not, the question is how to design our interface so that we can do the things `gather` can do, **as well as some things which it cannot do**, safely and with minimal overhead in both typing and CPU.

**Comment by Sxderp** [comment]
> as well as some things which it cannot do

I don't understand why that's a goal in this situation.

**Comment by m-u-xyz** [comment]
> I don't understand why that's a goal in this situation.

Maybe because a well-thought-out interface that happens to cover as many usecases as are reasonable (while not covering usecases that are not, most notably the ability to leave unaccounted-for tasks by mistake **or** design) is ultimately more useful than "hey asyncio has this `gather` thing that people are used to, so why don't we add that". 

Let's face it, asyncio is a straight evolution of "look we hacked some concurrency into Python with that new `yield` thing, oh there's now a `yield from`, oh we can add some better syntax around that, oh hey we can wrap a coroutine in a future let's call that thing a task, oh let's add some way to wait for more than one of these at a time". Like real-life evolution this ends up in a dead end or two sometime, which is why Trio (and anyio) exist in the first place.

Sorry if I'm a bit opinionated about this, but I've commited my share of unfixable interface design mistakes …
